src/Inventory.zig

```diff
@@ -1555,9 +1556,82 @@ pub const Command = struct { // MARK: Command
 	const UpdateBlock = struct { // MARK: UpdateBlock
 		source: InventoryAndSlot,
 		pos: Vec3i,
+		dropLocation: BlockDropLocation,
 		oldBlock: Block,
 		newBlock: Block,
 
+		const half = @as(Vec3f, @splat(0.5));
+		const itemHitBoxMargin: f32 = @floatCast(main.itemdrop.ItemDropManager.radius);
+		const itemHitBoxMarginVec = @as(Vec3f, @splat(itemHitBoxMargin));
+
+		const BlockDropLocation = struct {
+			dir: Neighbor,
+			min: Vec3f,
+			max: Vec3f,
+
+			pub fn drop(self: BlockDropLocation, pos: Vec3i, newBlock: Block, _drop: main.blocks.BlockDrop) void {
+				if(newBlock.collide()) {
+					self.dropOutside(pos, _drop);
+				} else {
+					self.dropInside(pos, _drop);
+				}
+			}
+			fn dropInside(self: BlockDropLocation, pos: Vec3i, _drop: main.blocks.BlockDrop) void {
+				for(_drop.items) |itemStack| {
+					main.server.world.?.drop(itemStack.clone(), self.insidePos(pos), self.dropDir(), self.dropVelocity());
+				}
+			}
+			fn insidePos(self: BlockDropLocation, _pos: Vec3i) Vec3d {
+				const pos: Vec3d = @floatFromInt(_pos);
+				return pos + self.randomOffset();
+			}
+			fn randomOffset(self: BlockDropLocation) Vec3f {
+				const max: Vec3f = @min(@as(Vec3f, @splat(1.0)) - itemHitBoxMarginVec, @max(itemHitBoxMarginVec, self.max - itemHitBoxMarginVec));
+				const min: Vec3f = @min(max, @max(itemHitBoxMarginVec, self.min + itemHitBoxMarginVec));
+				const center: Vec3f = (max + min)*half;
+				const width: Vec3f = (max - min)*half;
+				const factors: Vec3f = main.random.nextFloatVectorSigned(3, &main.seed)*half;
+				const offset: Vec3f = center + width*factors;
```

Comment:
I'm sure most of these don't need the explicit type declaration.

Reply:
But is that harmful, that they have names?

Reply:
Anyway, I have removed factors variable as it didn't provide any value, but I would prefer to keep the rest.

Reply:
Well, but adding the types everywhere doesn't make it clearer to read, the function is short, the inputs and outputs are Vec3f, why write this everywhere?

Reply:
> to ensure all the math is done with the precision I want

The return type should be enough for this promise. Zig will never narrow from f64 to f32.

Reply:
Ok, so I stripped the type annotations, is it enough or do you insist on removing the variables?